### ISSUE TITLE
Missing tracing for AWS MySQL driver

### ISSUE THREAD

**Issue by ondrejfuhrer** [DataDog/dd-trace-java#3443]
AWS has their on MySQL image with some AWS specific features. However after using the driver the database tracking doesn't work properly.
Official doc: https://awslabs.github.io/aws-mysql-jdbc/
GitHub: https://github.com/awslabs/aws-mysql-jdbc

My assumption is that the driver is missing from https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/instrumentation/jdbc/src/main/java/datadog/trace/instrumentation/jdbc/ConnectionInstrumentation.java#L11 list, but not sure if adding it is enough 🙂

**Comment by devinsba** [comment]
Thanks for the report. We'll also have to add their PreparedStatement classes as well.

You can work around this until we have a release with these included by modifying the `dd.trace.jdbc.connection.class.name` and `dd.trace.jdbc.prepared.statement.class.name` system properties, or the corresponding environment variables. (`DD_TRACE_JDBC_CONNECTION_CLASS_NAME` and `DD_TRACE_JDBC_PREPARED_STATEMENT_CLASS_NAME`)

**Comment by ondrejfuhrer** [comment]
Hey @devinsba 

Sure, this is the connection string that we tried - 
```
dataSourceBuilder.url("jdbc:mysql:aws://" + dbHost + ":" + dbPort + "/" + dbName)
```

The driver was set to `software.aws.rds.jdbc.mysql.Driver`.
Let me know if you need any further info.

**Comment by ondrejfuhrer** [comment]
Hey @devinsba , was there any progress on this issue?

**Comment by devinsba** [comment]
I've found a couple bugs that might be preventing this from working. I'm mostly focused on a different task at the moment I'll let you know if I make any progress

**Comment by maksimsc** [comment]
Hey
@devinsba just to clarify, currently tracing is not working properly with the new aws MySQL driver? It should theoretically be supported as the connection type is the same, although it is not working, right?

At least I've also tried default setup and explicit configuration for `DD_TRACE_JDBC_CONNECTION_CLASS_NAME` and `DD_TRACE_JDBC_PREPARED_STATEMENT_CLASS_NAME` and was not able to see the db traces.

**Comment by devinsba** [comment]
We found one issue already while building a reproducing test. (#3502) I'll try to working on this a bit each day and post any progress here

**Comment by maksimsc** [comment]
@devinsba thank you. Let me know if I can help somehow – testing or something.

**Comment by ondrejfuhrer** [comment]
Hey @devinsba , I noticed that your fix has been released in `0.100.0`, great! But also in the PR you state that this is not yet all that needs to be done, is that correct? Just wanna confirm if it makes sense to test it again on our end 🙂

**Comment by devinsba** [comment]
Correct. It was the first bug I found while debugging that actually prevented me from getting further. I'm trying to adapt our tests to support this driver currently so I can see what else is broken

**Comment by maksimsc** [comment]
Hey, @devinsba Any update here?

**Comment by devinsba** [comment]
I ran into some issues while testing, it turns out the driver only works when connected to an RDS instance because of some server metadata queries that are run during connection. I am working on getting an environment for integration testing with the driver

**Comment by Loganathanc** [comment]
i do not know how to create PR and unable to do it. however here is the changes and got worked.

Replace the methods by given method in below.

**AbstractConnectionInstrumentation.java**

@Override
  public Map<String, String> contextStore() {
    _//return singletonMap("java.sql.Statement", DBQueryInfo.class.getName());_ 
    return singletonMap("java.sql.PreparedStatement", DBQueryInfo.class.getName());
  }

public static class ConnectionPrepareAdvice {
    @Advice.OnMethodExit(suppress = Throwable.class)
    public static void addDBInfo(
        @Advice.This Connection connection,
        @Advice.Argument(0) final String sql,
        @Advice.Return final PreparedStatement statement) {
      _//ContextStore<Statement, DBQueryInfo> contextStore =
       //  InstrumentationContext.get(Statement.class, DBQueryInfo.class);_
      ContextStore<PreparedStatement, DBQueryInfo> contextStore =
          InstrumentationContext.get(PreparedStatement.class, DBQueryInfo.class);
      if (null == contextStore.get(statement)) {
        DBQueryInfo info = DBQueryInfo.ofPreparedStatement(sql);
        contextStore.put(statement, info);
        logQueryInfoInjection(connection, statement, info);
      }
    }


**AbstractPreparedStatementInstrumentation**

import datadog.trace.bootstrap.ContextStore;
import java.sql.PreparedStatement;

 @Override
  public Map<String, String> contextStore() {
    Map<String, String> contextStore = new HashMap<>(4);
    _//contextStore.put("java.sql.Statement", DBQueryInfo.class.getName());_
    contextStore.put("java.sql.PreparedStatement", DBQueryInfo.class.getName());
    contextStore.put("java.sql.Statement", Boolean.class.getName());
    contextStore.put("java.sql.Connection", DBInfo.class.getName());
    return contextStore;
  }

public static class PreparedStatementAdvice {

    @Advice.OnMethodEnter(suppress = Throwable.class)
    _//public static AgentScope onEnter(@Advice.This final Statement statement) {
     // int depth = CallDepthThreadLocalMap.incrementCallDepth(Statement.class);
     // if (depth > 0) {_
    public static AgentScope onEnter(@Advice.This final PreparedStatement statement) {
      ContextStore<Statement, Boolean> interceptionTracker =
          InstrumentationContext.get(Statement.class, Boolean.class);
      if (Boolean.TRUE.equals(interceptionTracker.get(statement))) {
        return null;
      }
      interceptionTracker.put(statement, Boolean.TRUE);
      try {
        Connection connection = statement.getConnection();
        DBQueryInfo queryInfo =
            _//InstrumentationContext.get(Statement.class, DBQueryInfo.class).get(statement);_
            InstrumentationContext.get(PreparedStatement.class, DBQueryInfo.class).get(statement);
        if (null == queryInfo) {
          logMissingQueryInfo(statement);
          return null;
        }

        final AgentSpan span = startSpan(DATABASE_QUERY);
        DECORATE.afterStart(span);
        DECORATE.onConnection(
            span, connection, InstrumentationContext.get(Connection.class, DBInfo.class));
        DECORATE.onPreparedStatement(span, queryInfo);
        return activateSpan(span);
      } catch (SQLException e) {
        logSQLException(e);
        // if we can't get the connection for any reason
        return null;
      }
    }

    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
    public static void stopSpan(
        _//@Advice.Enter final AgentScope scope, @Advice.Thrown final Throwable throwable) {_
        @Advice.This PreparedStatement statement,
        @Advice.Enter final AgentScope scope,
        @Advice.Thrown final Throwable throwable) {
      if (scope == null) {
        return;
      }
      DECORATE.onError(scope.span(), throwable);
      DECORATE.beforeFinish(scope.span());
      scope.close();
      scope.span().finish();
      _//CallDepthThreadLocalMap.reset(Statement.class);_
      InstrumentationContext.get(PreparedStatement.class, DBQueryInfo.class).put(statement, null);
    }
  }

**ConnectionInstrumentation.java**

      // AWS
      ,"software.aws.rds.jdbc.mysql.shading.com.mysql.cj.jdbc.ConnectionImpl",


**PreparedStatementInstrumentation**

 //aws
      "software.aws.rds.jdbc.mysql.shading.com.mysql.cj.jdbc.ClientPreparedStatement",
      "software.aws.rds.jdbc.mysql.shading.com.mysql.cj.jdbc.ServerPreparedStatement",
      "software.aws.rds.jdbc.mysql.shading.com.mysql.cj.jdbc.CallableStatement",


These 4 files if modify as per the above code, you will get the traces in Datadog. its working for me.

**Comment by devinsba** [comment]
I don't know if all of those changes are required. This change might be enough now https://github.com/DataDog/dd-trace-java/pull/3580

Snapshot here: https://output.circle-artifacts.com/output/job/0c4ced80-4d7c-4d48-bcfb-b8db7e0a3877/artifacts/0/libs/dd-java-agent-0.103.0-SNAPSHOT.jar

**Comment by Loganathanc** [comment]
Yes. its working fine. Please push the changes as part of upcoming release. thanks
